src/app/clusters/commissioner-control-server/commissioner-control-server.cpp

```diff
@@ -0,0 +1,277 @@
+{
+    EventNumber eventNumber;
+    CHIP_ERROR error = LogEvent(result, kRootEndpointId, eventNumber);
+    if (CHIP_NO_ERROR != error)
+    {
+        ChipLogError(Zcl, "CommissionerControl: Unable to emit CommissioningRequestResult event: %s", error.AsString());
+    }
+
+    return error;
+}
+
+} // namespace CommissionerControl
+} // namespace Clusters
+} // namespace app
+} // namespace chip
+
+bool emberAfCommissionerControlClusterRequestCommissioningApprovalCallback(
+    app::CommandHandler * commandObj, const app::ConcreteCommandPath & commandPath,
+    const Clusters::CommissionerControl::Commands::RequestCommissioningApproval::DecodableType & commandData)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    Status status  = Status::Success;
+
+    ChipLogProgress(Zcl, "Received command to request commissioning approval");
+
+    auto sourceNodeId = getNodeId(commandObj);
+
+    // Check if the command is executed via a CASE session
+    if (sourceNodeId == kUndefinedNodeId)
+    {
+        ChipLogError(Zcl, "Command not executed via CASE session, failing with UNSUPPORTED_ACCESS");
+        commandObj->AddStatus(commandPath, Status::UnsupportedAccess);
+        return true;
+    }
+
+    auto fabricIndex = commandObj->GetAccessingFabricIndex();
+    auto requestId   = commandData.requestId;
+    auto vendorId    = commandData.vendorId;
+    auto productId   = commandData.productId;
+
+    // The label assigned from commandData need to be stored in CommissionerControl::Delegate which ensure that the backing buffer
+    // of it has a valid lifespan during fabric sync setup process.
+    auto & label = commandData.label;
+
+    // Create a CommissioningApprovalRequest struct and populate it with the command data
+    Clusters::CommissionerControl::CommissioningApprovalRequest request = { .requestId    = requestId,
+                                                                            .vendorId     = vendorId,
+                                                                            .productId    = productId,
+                                                                            .clientNodeId = sourceNodeId,
+                                                                            .fabricIndex  = fabricIndex,
+                                                                            .label        = label };
+
+    Clusters::CommissionerControl::Delegate * delegate =
+        Clusters::CommissionerControl::CommissionerControlServer::Instance().GetDelegate();
+
+    VerifyOrExit(delegate != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+
+    // Handle commissioning approval request
+    err = delegate->HandleCommissioningApprovalRequest(request);
+
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(Zcl, "emberAfCommissionerControlClusterRequestCommissioningApprovalCallback error: %s", err.AsString());
+        status = Status::Failure;
```

Comment:
I strongly suggest converting the actual error to a status, in case the delegate wants to communicate a specific status.:
```suggestion
        status = StatusIB(err).mStatus;
```

or so.